spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/core/PubSubOperations.java

```diff
@@ -16,14 +16,39 @@
 
 package org.springframework.cloud.gcp.pubsub.core;
 
+import com.google.cloud.pubsub.v1.MessageReceiver;
+import com.google.cloud.pubsub.v1.Subscriber;
+
 import org.springframework.messaging.Message;
 import org.springframework.util.concurrent.ListenableFuture;
 
 /**
+ * An abstraction for Google Cloud Pub/Sub.
+ *
  * @author Vinicius Carvalho
+ * @author João André Martins
  */
 public interface PubSubOperations {
 
-	ListenableFuture<String> send(String topic, Message message);
+	/**
+	 * Sends a Spring {@link Message} to Pub/Sub.
+	 *
+	 * @param topic the name of an existing topic
+	 * @param message the Spring message
+	 * @return the listenable future of the call
+	 */
+	ListenableFuture<String> publish(String topic, Message message);
+
+	/**
+	 * Adds a callback method to an existing subscription.
+	 *
+	 * <p>
+	 * The created {@link Subscriber} is returned so it can be stopped.
+	 *
+	 * @param subscription the name of an existing subscription
+	 * @param messageHandler the callback method triggered when new messages arrive
+	 * @return subscriber listening to new messages
+	 */
+	Subscriber subscribe(String subscription, MessageReceiver messageHandler);
```

Comment:
I'd like to make a suggestion here: instead of a `MessageReceiver`, how about we provide our own callback format?

The reason is that `MessageReceiver` makes users handle a `PubsubMessage`, which is undesirable I think.

In `PubSubTemplate`, we could translate from a `PubsubMessage` to a `Message` and let then call our own callback on that.

Thoughts before I jump into implementing this? cc @saturnism

Reply:
Agreed, either that, or change the `publish` method to accept a `PubsubMessage` as well, instead of `Message` to keep the API consistent.

Reply:
does it make sense to just use the `org.springframework.messaging.MessageHandler` then?

Reply:
If you would like to go this way, I would suggest to have it as variant. Don't forget that we have `PubSubInboundChannelAdapter`, which looks like use this `PubSubOperations` and that would be much better to let it to deal with the low level protocol directly: there are many angles to be implemented exactly in the Channel Adapter, like headers mapping and some additional headers indicating that we are exactly from the Channel Adapter.

I somehow last time don't have strong opinion about APIs: if people made their vote then they had had some reason and if it really makes some sense I don't mind. Like I said about this `poll()` before. Although IMO this subscribe and process is not an operation of the template: it really must be stateless. That's why `RabbitTemplate` and `JmsTemplate` don't have long-living consuming operations. That's already a `MessageListenerContainer` responsibility. But you see: I still don't mind. Just because those two pretty old (but stable) API, and here we introduce a new modern API, so I'm fully open to give it a shot and let people to deal with callback API from our template. But on the other hand I would keep in mind `PubSubInboundChannelAdapter` and wouldn't try to eliminate its possible flexibility.

So, feel free to mix Messaging API in this template (although in other places we have it as a separate one - `RabbitMessagingTemplate`), but, please, keep the raw API to be able to manipulate with it from other places.

Reply:
@markfisher that `MessageHandler` seems just like what I was thinking. Thanks!

@artembilan the main reason I wanted to do this is exactly message conversion, and I came across this issue exactly on `PubSubInboundChannelAdapter`! The issue is `PubSubTemplate` contains the `MessageConverter` to turn `Messages` into payload and vice-versa. By giving a `PubsubMessage` to `PubSubInboundChannelAdapter`, we make it also have to know about message converters, which is probably avoidable.

I'll keep your suggestion in mind and keep the low-level trigger. We can revisit this later if needed be.

Reply:
I realise that Messaging is only really needed for the channel adapters, so we might want to keep `PubSubOperations` Spring Messaging-agnostic, or provide utility Messaging methods like suggested.
I guess the issue I should be trying to tackle in `PubSubTemplate` is payload conversion (Object2Jackson/Gson and vice-versa). We might even be able to move the `MessageConverter` out of `PubSubTemplate` and into the adapters.